mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala

```diff
@@ -0,0 +1,290 @@
+    * @group param
+    */
+  val strategy: Param[String] = new Param(this, "strategy", "strategy for imputation. " +
+    "If mean, then replace missing values using the mean value of the feature." +
+    "If median, then replace missing values using the median value of the feature." +
+    "If most, then replace missing using the most frequent value of the feature.",
+    ParamValidators.inArray[String](Imputer.supportedStrategyNames.toArray))
+
+  /** @group getParam */
+  def getStrategy: String = $(strategy)
+
+  /**
+    * The placeholder for the missing values. All occurrences of missingValue will be imputed.
+    * Default: Double.NaN
+    *
+    * @group param
+    */
+  val missingValue: DoubleParam = new DoubleParam(this, "missingValue",
+    "The placeholder for the missing values. All occurrences of missingValue will be imputed")
+
+  /** @group getParam */
+  def getMissingValue: Double = $(missingValue)
+
+  /** Validates and transforms the input schema. */
+  protected def validateAndTransformSchema(schema: StructType): StructType = {
+    validateParams()
+    val inputType = schema($(inputCol)).dataType
+    require(inputType.isInstanceOf[VectorUDT] || inputType.isInstanceOf[DoubleType],
+      s"Input column ${$(inputCol)} must of type Vector or Double")
+    require(!schema.fieldNames.contains($(outputCol)),
+      s"Output column ${$(outputCol)} already exists.")
+    val outputFields = schema.fields :+ StructField($(outputCol), new VectorUDT, false)
+    StructType(outputFields)
+  }
+
+}
+
+/**
+ * :: Experimental ::
+ * Imputation estimator for completing missing values, either using the mean, the median or
+ * the most frequent value of the column in which the missing values are located. This class
+ * also allows for different missing values.
+ */
+@Experimental
+class Imputer @Since("2.0.0")(override val uid: String)
+  extends Estimator[ImputerModel] with ImputerParams with DefaultParamsWritable {
+
+  @Since("2.0.0")
+  def this() = this(Identifiable.randomUID("imputer"))
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Imputation strategy. Available options are "mean", "median" and "most".
+   * @group setParam
+   */
+  def setStrategy(value: String): this.type = set(strategy, value)
+
+  /** @group setParam */
+  def setMissingValue(value: Double): this.type = set(missingValue, value)
+
+  setDefault(strategy -> "mean", missingValue -> Double.NaN)
+
+  override def fit(dataset: DataFrame): ImputerModel = {
+    val alternate = dataset.select($(inputCol)).schema.fields(0).dataType match {
+      case DoubleType =>
+        Vectors.dense(getColStatistics(dataset, $(inputCol)))
+      case _: VectorUDT =>
+        val vl = dataset.first().getAs[Vector]($(inputCol)).size
+        val statisticsArray = new Array[Double](vl)
+        (0 until vl).foreach(i => {
```

Comment:
This could be a big performance issue with large vectors, as we could be running 100s (or millions!) of SQL queries sequentially... For vectors I favour the `colStats` approach of using `MultivariateOnlineSummarizer` for efficiency. I think that if we support vectors here, we should rather enable ignoring the NaNs in `colStats` or a deeper API (even if we do it as `private` or `DevelopApi` for now).